bip-0325.mediawiki

```diff
@@ -26,36 +26,37 @@ A new type of test network would be more suitable for integration testing by org
 
 A new type of network ("signet"), which takes an additional consensus parameter called the challenge (scriptPubKey). The challenge can be a simple pubkey (P2PKH style), or a k-of-n multisig, or any other script you would want.
 
-The witness commitment of the coinbase transaction is extended to include a secondary commitment (the signature/solution):
+The witness commitment of the coinbase transaction is extended to include a secondary commitment (the signature/solution) of either:
 
     1-4 bytes - Push the following (x + 4) bytes
-    4 bytes - Signet header (0xecc7daa2)
-    x bytes - Solution (sigScript)
-
-Any push operations that do not start with the 4 byte signet header are ignored. Multiple push operations with the 4 byte signet header are ignored except for the first entry.
-
-Any signature operations contained within the challenge use SHA256d(modifiedBlockHash), i.e. the double-SHA256 digest of the following data as the sighash:
-
-{|class="wikitable" style="text-align: center;"
-|-
-!Type
-!Size
-!Name
-|-
-|Int32||4||nVersion
-|-
-|Uint256||32||hashPrevBlock
-|-
-|Uint256||32||modifiedMerkleRoot
-|-
-|Uint32||4||nTime
-|-
-|Uint32||4||nBits
-|}
-
-The <code>modifiedMerkleRoot</code> hash is obtained by generating the merkle root of the block transactions, with the coinbase witness commitment as is, without the signet extension. This means the merkle root of the block is different from the merkle root in the signet commitment. This is needed, because the signature can never be included in the very message (in this case, a block) that is being signed. Apart from the signature, to facilitate block generation (mining), the block nonce value is the only other component of the block that the signet signature does not commit to. When grinding proof of work, the extended nonce cannot be used as it would invalidate the signature. Instead, simply resigning the same (or an updated) block will give a new search space.
-
-A block is considered fully validated if the above commitment is found, and its solution is valid. It is recommended that this verification is done directly before or after the witness commitment verification, as the data required to do both is approximately the same.
+    4 bytes - Signet scriptSig header (0xecc7daa2)
+    x bytes - Solution (scriptSig)
+
+or
+
+    1-4 bytes - Push the following (x + 4) bytes
+    4 bytes - Signet scriptWitness header (0xecc7daa3)
+    x bytes - Solution (scriptWitness)
+
+Any push operations that do not start with either of the 4 byte signet header are ignored. Multiple push operations with either 4 byte signet header are ignored except for the first instance of each header.
+
+To sign the block or verify a block signature, a virtual transaction with a single input and output is constructed from the block as follows:
+
+    nVersion = 1
+    nLockTime = block.nTime
+    vin[0].prevout.hash = block.hashPrevBlock
```

Comment:
I'm not sure if there were perhaps reasons not to do this, but maybe it is desirable to exactly structure it like transaction spending, as that would mean PSBT infrastructure can be reused for block signing.

One possibility would be, construct 2 virtual transactions. The first is a coinbase tx (perhaps intentionally BIP34-invalid), with a single output, whose scriptPubKey is the signet's challenge script. The second tx spends that output, referring to it using its txid, and its scriptSig/witness is what is put in the block. The first tx could even be cached and constant for all blocks.

Reply:
PSBT compatibility would be great, yeah!

Reply:
I've updated the text and the sample implementation to use this approach, it seems much better!

One reason we might not have thought of it, is that an idea had been to allow signet users to opt-in to reorgs by allowing a different challenge: ie, normally signet might be checked against "2 KEY_A KEY_B 2 CHECKMULTISIG" and the solution is "0 SIG_A SIG_B"; but if you were to use "DROP 1 KEY_A 1 CHECKMULTISIG" as your script, and KEY_A alone was used to sign short reorgs as well as the main chain, you'd be able to test how well your software handled reorgs, while people who used the normal signet challenge wouldn't see any of the reorgs. By actually constructing a tx to spend that includes the challenge, this doesn't work so well anymore.

On the other hand, it probably doesn't work anyway, because at the p2p layer, the peer doing the reorgs would try forwarding the to-be-reorged blocks to its peers, who would see them as consensus-invalid and ban the peer. So I think it's fine to drop this idea in favour of PSBT support.

Reply:
aside: I think it's kind of awesome you could get a ColdCard, put it into HSM mode, and have it blocksign!

Reply:
I never thought of actually having multiple networks with the same underlying blockchain. That should totally work -- unfortunately signet now derives the magic from the block challenge, so you can't tweak that easily (well, you can, but it becomes a little messy). Perhaps add a -signet_magic flag that lets you override the derived one..

Reply:
I think you have to have a reindex if you're applying a different reorg policy -- otherwise, suppose your chain goes 100a 101a 102a 103a 104b 105b (for height 100-105) but the "b" blocks will eventually be reorged out and replaced by 104a 105a 106a but you've only seen 104a so far and that's a less-work tip. Switching to "no reorgs" means you want to move to 104a as your tip immediately. I guess you could do that with a specialised rewind instead of a reindex?

Reply:
As the challenge can only be changed at startup, simply disconnecting blocks until the just removed block can be connected again should suffice.